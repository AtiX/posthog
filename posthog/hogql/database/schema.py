from pydantic import BaseModel, Extra

from posthog.hogql.database.joins import join_with_max_person_distinct_id_table, join_with_persons_table
from posthog.hogql.database.primitives import (
    BooleanDatabaseField,
    DateTimeDatabaseField,
    FieldTraverser,
    IntegerDatabaseField,
    LazyTable,
    StringDatabaseField,
    StringJSONDatabaseField,
    Table,
    VirtualTable,
)


class EventsPersonSubTable(VirtualTable):
    id: StringDatabaseField = StringDatabaseField(name="person_id")
    created_at: DateTimeDatabaseField = DateTimeDatabaseField(name="person_created_at")
    properties: StringJSONDatabaseField = StringJSONDatabaseField(name="person_properties")

    def clickhouse_table(self):
        return "events"


class PersonsTable(Table):
    id: StringDatabaseField = StringDatabaseField(name="id")
    created_at: DateTimeDatabaseField = DateTimeDatabaseField(name="created_at")
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
    properties: StringJSONDatabaseField = StringJSONDatabaseField(name="properties")
    is_identified: BooleanDatabaseField = BooleanDatabaseField(name="is_identified")
    is_deleted: BooleanDatabaseField = BooleanDatabaseField(name="is_deleted")
    version: IntegerDatabaseField = IntegerDatabaseField(name="version")

    def clickhouse_table(self):
        return "person"


class PersonStaticCohortTable(Table):
    id: StringDatabaseField = StringDatabaseField(name="id")
    person_id: StringDatabaseField = StringDatabaseField(name="person_id")
    cohort_id: IntegerDatabaseField = IntegerDatabaseField(name="cohort_id")
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")

    person: LazyTable = LazyTable(from_field="person_id", table=PersonsTable(), join_function=join_with_persons_table)

    def clickhouse_table(self):
        return "person_static_cohort"


class CohortPeopleTable(Table):
    person_id: StringDatabaseField = StringDatabaseField(name="person_id")
    cohort_id: IntegerDatabaseField = IntegerDatabaseField(name="cohort_id")
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
    sign: IntegerDatabaseField = IntegerDatabaseField(name="sign")
    version: IntegerDatabaseField = IntegerDatabaseField(name="version")

    person: LazyTable = LazyTable(from_field="person_id", table=PersonsTable(), join_function=join_with_persons_table)

    def clickhouse_table(self):
        return "cohortpeople"


class PersonDistinctIdTable(Table):
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
    distinct_id: StringDatabaseField = StringDatabaseField(name="distinct_id")
    person_id: StringDatabaseField = StringDatabaseField(name="person_id")
    is_deleted: BooleanDatabaseField = BooleanDatabaseField(name="is_deleted")
    version: IntegerDatabaseField = IntegerDatabaseField(name="version")

    person: LazyTable = LazyTable(from_field="person_id", table=PersonsTable(), join_function=join_with_persons_table)

    # static_cohorts: LazyTable(from_field="person_id", table=)

    def avoid_asterisk_fields(self):
        return ["is_deleted", "version"]

    def clickhouse_table(self):
        return "person_distinct_id2"


class EventsTable(Table):
    uuid: StringDatabaseField = StringDatabaseField(name="uuid")
    event: StringDatabaseField = StringDatabaseField(name="event")
    properties: StringJSONDatabaseField = StringJSONDatabaseField(name="properties")
    timestamp: DateTimeDatabaseField = DateTimeDatabaseField(name="timestamp")
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
    distinct_id: StringDatabaseField = StringDatabaseField(name="distinct_id")
    elements_chain: StringDatabaseField = StringDatabaseField(name="elements_chain")
    created_at: DateTimeDatabaseField = DateTimeDatabaseField(name="created_at")

    # lazy table that adds a join to the persons table
    pdi: LazyTable = LazyTable(
        from_field="distinct_id", table=PersonDistinctIdTable(), join_function=join_with_max_person_distinct_id_table
    )
    # person fields on the event itself
    poe: EventsPersonSubTable = EventsPersonSubTable()

    # TODO: swap these between pdi and person_on_events as needed
    person: FieldTraverser = FieldTraverser(chain=["pdi", "person"])
    person_id: FieldTraverser = FieldTraverser(chain=["pdi", "person_id"])

    def clickhouse_table(self):
        return "events"


class SessionRecordingEvents(Table):
    uuid: StringDatabaseField = StringDatabaseField(name="uuid")
    timestamp: DateTimeDatabaseField = DateTimeDatabaseField(name="timestamp")
    team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
    distinct_id: StringDatabaseField = StringDatabaseField(name="distinct_id")
    session_id: StringDatabaseField = StringDatabaseField(name="session_id")
    window_id: StringDatabaseField = StringDatabaseField(name="window_id")
    snapshot_data: StringJSONDatabaseField = StringJSONDatabaseField(name="snapshot_data")
    created_at: DateTimeDatabaseField = DateTimeDatabaseField(name="created_at")
    has_full_snapshot: BooleanDatabaseField = BooleanDatabaseField(name="has_full_snapshot")
    events_summary: StringJSONDatabaseField = StringJSONDatabaseField(name="events_summary", array=True)
    click_count: IntegerDatabaseField = IntegerDatabaseField(name="click_count")
    keypress_count: IntegerDatabaseField = IntegerDatabaseField(name="keypress_count")
    timestamps_summary: DateTimeDatabaseField = DateTimeDatabaseField(name="timestamps_summary", array=True)
    first_event_timestamp: DateTimeDatabaseField = DateTimeDatabaseField(name="first_event_timestamp")
    last_event_timestamp: DateTimeDatabaseField = DateTimeDatabaseField(name="last_event_timestamp")
    urls: StringDatabaseField = StringDatabaseField(name="urls", array=True)

    pdi: LazyTable = LazyTable(
        from_field="distinct_id", table=PersonDistinctIdTable(), join_function=join_with_max_person_distinct_id_table
    )

    def clickhouse_table(self):
        return "session_recording_events"


class Database(BaseModel):
    class Config:
        extra = Extra.forbid

    # Users can query from the tables below
    events: EventsTable = EventsTable()
    persons: PersonsTable = PersonsTable()
    person_distinct_ids: PersonDistinctIdTable = PersonDistinctIdTable()
    session_recording_events: SessionRecordingEvents = SessionRecordingEvents()

    def has_table(self, table_name: str) -> bool:
        return hasattr(self, table_name)

    def get_table(self, table_name: str) -> Table:
        if self.has_table(table_name):
            return getattr(self, table_name)
        raise ValueError(f'Table "{table_name}" not found in database')


database = Database()
