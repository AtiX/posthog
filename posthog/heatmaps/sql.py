from django.conf import settings

from posthog.clickhouse.kafka_engine import kafka_engine
from posthog.clickhouse.table_engines import (
    Distributed,
    ReplicationScheme,
    MergeTreeEngine,
)
from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_HEATMAP_EVENTS

HEATMAPS_DATA_TABLE = lambda: "sharded_heatmaps"


"""
We intend to send specific $heatmap events to build a heatmap instead of building from autocapture like the click map
We'll be storing individual clicks per url/team/session
And we'll be querying for those clicks at day level of granularity
And we'll be querying by URL exact or wildcard match
And we'll _sometimes_ be querying by width

We _could_ aggregate this data by day, but we're hoping this will be small/fast enough not to bother
And can always add a materialized view for day (and week?) granularity driven by this data if needed

We only add session_id so that we could offer example sessions for particular clicked areas in the toolbar
"""

KAFKA_HEATMAPS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    $session_id VARCHAR,
    team_id Int64,
    timestamp DateTime64(6, 'UTC'),
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x Int16,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y Int16,
    -- stored so that in future we can support other resolutions
    scale_factor Int16,
    $viewport_width Int16,
    $viewport_height Int16,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed Bool,
    $current_url VARCHAR
) ENGINE = {engine}
"""

HEATMAPS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    $session_id VARCHAR,
    team_id Int64,
    timestamp DateTime64(6, 'UTC'),
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x Int16,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y Int16,
    -- stored so that in future we can support other resolutions
    scale_factor Int16,
    $viewport_width Int16,
    $viewport_height Int16,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed Bool,
    $current_url VARCHAR,
    _timestamp DateTime
) ENGINE = {engine}
"""

HEATMAPS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("heatmaps", replication_scheme=ReplicationScheme.SHARDED)

HEATMAPS_TABLE_SQL = lambda: (
    HEATMAPS_TABLE_BASE_SQL
    + """
    PARTITION BY toYYYYMM(timestamp)
    -- almost always this is being queried by
    --   * team_id,
    --   * date range,
    --   * URL (maybe matching wild cards),
    --   * width
    -- we'll almost never query this by session id
    -- so from least to most cardinality that's
    ORDER BY (team_id,  toDate(timestamp), $current_url, $viewport_width)
-- I am purposefully not setting index granularity
-- the default is 8192, and we will be loading a lot of data
-- per query, we tend to copy this 512 around the place but
-- i don't think it applies here
"""
).format(
    table_name=HEATMAPS_DATA_TABLE(),
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=HEATMAPS_DATA_TABLE_ENGINE(),
)

KAFKA_HEATMAPS_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
    table_name="kafka_heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS),
)

HEATMAPS_TABLE_MV_SQL = (
    lambda: """
CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_mv ON CLUSTER '{cluster}'
TO {database}.{target_table}
AS SELECT
    $session_id,
    team_id,
    timestamp,
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y,
    -- stored so that in future we can support other resolutions
    scale_factor,
    $viewport_width,
    $viewport_height,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed,
    $current_url
FROM {database}.kafka_heatmaps
""".format(
        target_table="writable_heatmaps",
        cluster=settings.CLICKHOUSE_CLUSTER,
        database=settings.CLICKHOUSE_DATABASE,
    )
)

# Distributed engine tables are only created if CLICKHOUSE_REPLICATED

# This table is responsible for writing to sharded_heatmaps based on a sharding key.
WRITABLE_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
    table_name="writable_heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=Distributed(
        data_table=HEATMAPS_DATA_TABLE(),
        # I don't think there's a great natural sharding key here
        # we'll be querying for team data by url and date,
        # so I _think_ this offers a reasonable spread of write load
        # without needing to query too many shards
        sharding_key="cityHash64(concat(toString(team_id), '-', $current_url, '-', toString(toDate(timestamp))))",
    ),
)

# This table is responsible for reading from heatmaps on a cluster setting
DISTRIBUTED_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
    table_name="heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=Distributed(
        data_table=HEATMAPS_DATA_TABLE(),
        # I don't think there's a great natural sharding key here
        # we'll be querying for team data by url and date,
        # so I _think_ this offers a reasonable spread of write load
        # without needing to query too many shards
        sharding_key="cityHash64(concat(toString(team_id), '-', $current_url, '-', toString(toDate(timestamp))))",
    ),
)

DROP_HEATMAPS_TABLE_SQL = lambda: (
    f"DROP TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
)

TRUNCATE_HEATMAPS_TABLE_SQL = lambda: (
    f"TRUNCATE TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
)
