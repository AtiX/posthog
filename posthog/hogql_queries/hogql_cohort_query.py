from typing import Any

from posthog.hogql_queries.actors_query_runner import ActorsQueryRunner
from posthog.schema import ActorsQuery
from posthog.queries.cohort_query import CohortQuery


class HogQLCohortQuery(CohortQuery):

    def get_query(self) -> tuple[str, dict[str, Any]]:
        if not self._outer_property_groups:
            # everything is pushed down, no behavioral stuff to do
            # thus, use personQuery directly

            # This just queries based on person properties and stuff
            # Need to figure out how to turn these cohort properties into a set of person properties
            actors_query = ActorsQuery(properties=[self._filter._data["properties"]])
            query_runner = ActorsQueryRunner(team=self._team, query=actors_query)
            return query_runner.to_query()