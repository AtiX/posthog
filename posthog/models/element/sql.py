CREATE_ELEMENTS_CHAIN_DAILY_COUNTS_TABLE = """
CREATE TABLE elements_chain_counts_daily
(
    day            DateTime64(3, 'UTC'),
    team_id        int,
    elements_chain String,
    "$current_url" String,
    count          int
)
ENGINE = SummingMergeTree
PARTITION BY toYYYYMM(day) ORDER BY (team_id,day,"$current_url", elements_chain);
"""

CREATE_ELEMENTS_CHAIN_DAILY_COUNTS_MV = """
CREATE MATERIALIZED VIEW elements_chain_counts_daily_mv
TO elements_chain_counts_daily
AS SELECT
    toStartOfDay(timestamp) as day,
    team_id,
    elements_chain,
    replaceRegexpAll(JSONExtractRaw(properties, '$current_url'), '^"|"$', '') as "$current_url",
    count(*) as count
FROM events
WHERE event = '$autocapture'
AND elements_chain != ''
GROUP BY team_id, day, elements_chain, "$current_url";
"""

GET_ELEMENTS_FROM_MV = """
SELECT
    elements_chain, sum(count) as count
FROM elements_chain_counts_daily_mv
WHERE
    team_id = %(team_id)s
    AND day >= toStartOfDay(%(date_from)s, 'UTC')
    AND day <= %(date_to)s
    AND "$current_url" = %(current_url)s
GROUP BY elements_chain
ORDER BY count DESC
"""


GET_ELEMENTS = """
SELECT
    elements_chain, count(1) as count
FROM events
WHERE
    team_id = %(team_id)s AND
    event = '$autocapture' AND
    elements_chain != ''
    {date_from}
    {date_to}
    {query}
GROUP BY elements_chain
ORDER BY count DESC
LIMIT {limit}{conditional_offset};
"""

GET_VALUES = """
SELECT
    extract(elements_chain, %(regex)s) as value, count(1) as count
FROM (
      SELECT elements_chain
      FROM events
      WHERE team_id = %(team_id)s
        AND event = '$autocapture'
        AND elements_chain != ''
        AND match(elements_chain, %(filter_regex)s)
        LIMIT 100000
)
GROUP BY value
ORDER BY count desc
LIMIT 100;
"""
