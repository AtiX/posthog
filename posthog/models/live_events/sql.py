from posthog.clickhouse.kafka_engine import kafka_engine
from posthog.clickhouse.table_engines import Distributed, MergeTreeEngine, ReplicationScheme
from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE, KAFKA_EVENTS_PLUGIN_INGESTION_TOPIC

LIVE_EVENTS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    uuid UUID,
    event VARCHAR,
    properties VARCHAR,
    timestamp DateTime64(6, 'UTC'),
    team_id Int64,
    distinct_id VARCHAR,
    elements_chain VARCHAR,
    created_at DateTime64(6, 'UTC'),
    _timestamp Nullable(DateTime)
) ENGINE = {engine}
"""


LIVE_EVENTS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine(
    "sharded_live_events", replication_scheme=ReplicationScheme.SHARDED
)

# TODO: Double check TTL and if we want to manually drop partitions e.g. every hour
LIVE_EVENTS_DATA_TABLE_SQL = lambda: (
    LIVE_EVENTS_TABLE_BASE_SQL
    + """PARTITION BY toStartOfTenMinutes(timestamp)
ORDER BY (team_id, timestamp, event, cityHash64(distinct_id), cityHash64(uuid))
TTL toDateTime(timestamp) + INTERVAL 1 DAY
"""
).format(
    table_name="sharded_live_events",
    cluster=CLICKHOUSE_CLUSTER,
    engine=LIVE_EVENTS_DATA_TABLE_ENGINE(),
)

# This table is responsible for writing to sharded_live_events based on a sharding key
# We also read from this table instead of sharded_live_events
DISTRIBUTED_LIVE_EVENTS_TABLE_SQL = lambda: LIVE_EVENTS_TABLE_BASE_SQL.format(
    table_name="live_events",
    cluster=CLICKHOUSE_CLUSTER,
    engine=Distributed(data_table="sharded_live_events", sharding_key="rand()"),  # use team_id?
)


KAFKA_LIVE_EVENTS_TABLE_SQL = lambda: LIVE_EVENTS_TABLE_BASE_SQL.format(
    table_name="kafka_live_events",
    cluster=CLICKHOUSE_CLUSTER,
    # being explicit about the group even though we're using the default
    # to highlight that this is important - we should never use the same
    # consumer group as the plugin server here!
    engine=kafka_engine(topic=KAFKA_EVENTS_PLUGIN_INGESTION_TOPIC, group="group1"),
)

# KLUDGE: We don't have the real timestamp yet, so we use now()
# currently this is handled in the plugin server through the
# handleTimestamp function, but we should port it to Django
#
# Note: we use 0 as created_at because the event hasn't been created yet,
# and that actually allows us to correctly aggregate arguments when
# doing a merge query
LIVE_EVENTS_MV_TABLE_SQL = lambda: """
CREATE MATERIALIZED VIEW live_events_mv ON CLUSTER '{cluster}'
TO {database}.{target_table}
AS SELECT
    uuid,
    event,
    properties,
    toDateTime64(now(), 6, 'UTC') as timestamp,
    team_id,
    distinct_id,
    '' as elements_chain,
    toDateTime64(0, 6, 'UTC') as created_at,
    _timestamp
FROM {database}.kafka_live_events
WHERE event != '$snapshot'
""".format(
    target_table="live_events",
    cluster=CLICKHOUSE_CLUSTER,
    database=CLICKHOUSE_DATABASE,
)


SELECT_LIVE_EVENTS_BY_TEAM_AND_CONDITIONS_FILTERS_SQL = """
SELECT
    uuid,
    event,
    team_id,
    distinct_id,
    properties_aggregated as properties,
    timestamp_aggregated as timestamp,
    elements_chain_aggregated as elements_chain,
    source_table
FROM (
    SELECT
        uuid,
        event,
        team_id,
        distinct_id,
        argMax(properties, created_at) as properties_aggregated,
        argMax(elements_chain, created_at) as elements_chain_aggregated,
        argMax(timestamp, created_at) as timestamp_aggregated,
        argMax(_table, created_at) as source_table
    FROM
        merge('{database}', '^(events|live_events)$')
    WHERE
        team_id = %(team_id)s
        {conditions}
        {filters}
    GROUP BY uuid, event, team_id, distinct_id
    ORDER BY timestamp_aggregated {order} {limit}
)
"""


SELECT_LIVE_EVENTS_BY_TEAM_AND_CONDITIONS_SQL = """
SELECT
    uuid,
    event,
    team_id,
    distinct_id,
    properties_aggregated as properties,
    timestamp_aggregated as timestamp,
    elements_chain_aggregated as elements_chain,
    source_table
FROM (
    SELECT
        uuid,
        event,
        team_id,
        distinct_id,
        argMax(properties, created_at) as properties_aggregated,
        argMax(elements_chain, created_at) as elements_chain_aggregated,
        argMax(timestamp, created_at) as timestamp_aggregated,
        argMax(_table, created_at) as source_table
    FROM
        merge('{database}', '^(events|live_events)$')
    WHERE
        team_id = %(team_id)s
        {conditions}
    GROUP BY uuid, event, team_id, distinct_id
    ORDER BY timestamp_aggregated {order} {limit}
)
"""

# test utilities
INSERT_INTO_LIVE_EVENTS_SQL = """
INSERT INTO live_events (
    uuid,
    event,
    properties,
    timestamp,
    team_id,
    distinct_id,
    elements_chain,
    created_at,
    _timestamp
)
VALUES (
    %(uuid)s,
    %(event)s,
    %(properties)s,
    %(timestamp)s,
    %(team_id)s,
    %(distinct_id)s,
    '',
    toDateTime64(0, 6, 'UTC'),
    now()
)
"""
