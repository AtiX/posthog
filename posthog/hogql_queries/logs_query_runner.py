from datetime import timedelta
from typing import cast
from dateutil.parser import isoparse
from django.utils.timezone import datetime

from posthog.clickhouse.client.connection import Workload
from posthog.hogql import ast
from posthog.hogql.parser import parse_expr, parse_select
from posthog.hogql.property import property_to_expr
from posthog.hogql_queries.insights.paginators import HogQLHasMorePaginator
from posthog.hogql_queries.query_runner import QueryRunner
from posthog.hogql_queries.utils.query_date_range import QueryDateRange
from posthog.models.filters.mixins.utils import cached_property
from posthog.schema import (
    CachedLogsQueryResponse,
    IntervalType,
    LogsQuery,
    LogsQueryResponse,
    LogsQueryResult,
)
from posthog.utils import relative_date_parse


class LogsQueryRunner(QueryRunner):
    query: LogsQuery
    response: LogsQueryResponse
    cached_response: CachedLogsQueryResponse

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.paginator = HogQLHasMorePaginator.from_limit_context(
            limit_context=self.limit_context, limit=self.query.limit, offset=self.query.offset
        )

    def to_query(self) -> ast.SelectQuery:
        return cast(
            ast.SelectQuery,
            parse_select(
                """
                    SELECT uuid, distinct_id, event, timestamp, properties.$level, properties.$msg, properties.$namespace, properties
                    FROM events
                    WHERE {where_clause}
                    ORDER BY timestamp DESC
                """,
                {"where_clause": self._where_clause()},
            ),
        )

    def to_actors_query(self) -> ast.SelectQuery:
        return self.to_query()

    def calculate(self) -> LogsQueryResponse:
        query = self.to_query()

        response = self.paginator.execute_hogql_query(
            query_type="LogsQuery",
            query=query,
            modifiers=self.query.modifiers or self.modifiers,
            team=self.team,
            workload=Workload.ONLINE,
            timings=self.timings,
            limit_context=self.limit_context,
        )

        results: list[LogsQueryResult] = [
            LogsQueryResult(
                uuid=str(log[0]),
                distinct_id=log[1],
                event=log[2],
                timestamp=str(log[3]),
                level=log[4],
                msg=log[5],
                namespace=log[6],
                properties=log[7],
            )
            for log in response.results
        ]

        return LogsQueryResponse(results=results, **self.paginator.response_params())

    def _where_clause(self):
        filters: list[ast.Expr] = []

        # Dates
        date_range_placeholders = self._query_date_range.to_placeholders()
        filters.extend(
            [
                parse_expr(
                    "timestamp >= {date_from_with_adjusted_start_of_interval}", placeholders=date_range_placeholders
                ),
                parse_expr("timestamp <= {date_to}", placeholders=date_range_placeholders),
            ]
        )

        # Before
        if self.query.before is not None:
            before = self.query.before
            try:
                parsed_date = isoparse(before)
            except ValueError:
                parsed_date = relative_date_parse(before, self.team.timezone_info)
            filters.append(
                parse_expr(
                    "timestamp < {timestamp}",
                    {"timestamp": ast.Constant(value=parsed_date)},
                    timings=self.timings,
                )
            )

        # After
        if self.query.after is not None:
            after = self.query.after
            if after != "all":
                try:
                    parsed_date = isoparse(after)
                except ValueError:
                    parsed_date = relative_date_parse(after, self.team.timezone_info)
                filters.append(
                    parse_expr(
                        "timestamp > {timestamp}",
                        {"timestamp": ast.Constant(value=parsed_date)},
                        timings=self.timings,
                    )
                )

        # Event name
        filters.append(parse_expr("event = '$log'"))

        # Search term
        if self.query.searchTerm is not None:
            filters.append(
                parse_expr(
                    """
                        or(
                            multiSearchAnyCaseInsensitive(properties.$msg, [{term}]) = 1,
                            multiSearchAnyCaseInsensitive(properties.$namespace, [{term}]) = 1
                        )
                    """,
                    {"term": ast.Constant(value=self.query.searchTerm)},
                )
            )

        # Properties
        if self.query.properties is not None and self.query.properties != []:
            filters.append(property_to_expr(self.query.properties, self.team))

        if len(filters) == 0:
            return ast.Constant(value=True)
        elif len(filters) == 1:
            return filters[0]

        return ast.And(exprs=filters)

    @cached_property
    def _query_date_range(self):
        return QueryDateRange(
            date_range=self.query.dateRange,
            team=self.team,
            interval=IntervalType.minute,
            now=datetime.now(),
        )

    def _is_stale(self, cached_result_package):
        return True

    def _refresh_frequency(self):
        return timedelta(minutes=1)
