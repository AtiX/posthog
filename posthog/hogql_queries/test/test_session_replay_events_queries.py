from posthog.clickhouse.client import sync_execute
from posthog.hogql_queries.hogql_query_runner import HogQLQueryRunner
from posthog.schema import HogQLQuery
from posthog.session_recordings.queries.test.session_replay_sql import produce_replay_summary
from posthog.session_recordings.sql.session_replay_event_sql import TRUNCATE_SESSION_REPLAY_EVENTS_TABLE_SQL
from posthog.test.base import ClickhouseTestMixin, APIBaseTest, QueryMatchingTest, snapshot_clickhouse_queries


class TestSessionReplayEventsHogQLQueries(ClickhouseTestMixin, APIBaseTest, QueryMatchingTest):
    def tearDown(self):
        sync_execute(TRUNCATE_SESSION_REPLAY_EVENTS_TABLE_SQL())

    def setUp(self):
        super().setUp()

        produce_replay_summary(
            team_id=self.team.id,
            session_id="session_id_one",
            distinct_id="distinct_id_one",
            mouse_activity_count=100,
        )
        produce_replay_summary(
            team_id=self.team.id,
            session_id="session_id_two",
            distinct_id="distinct_id_two",
            mouse_activity_count=200,
        )
        produce_replay_summary(
            team_id=self.team.id,
            session_id="session_id_two",
            distinct_id="distinct_id_two",
            mouse_activity_count=300,
        )

    @snapshot_clickhouse_queries
    def test_session_replay_events_table_does_not_need_to_be_grouped_by_session_id_explicitly(self):
        runner = HogQLQueryRunner(
            team=self.team,
            query=HogQLQuery(
                query="""
            select session_id, sum(mouse_activity_count) as mouse_activity_count
            from session_replay_events
            -- this table is always implicitly grouped by session id
            order by mouse_activity_count desc
            """
            ),
        )

        response = runner.calculate()
        assert response.results == [("session_id_two", 500), ("session_id_one", 100)]

    @snapshot_clickhouse_queries
    def test_session_replay_events_table_is_always_grouped_by_session_id(self):
        # test that the "not raw" table is always grouped by session id
        runner = HogQLQueryRunner(
            team=self.team,
            query=HogQLQuery(
                query="""
            select distinct_id, sum(mouse_activity_count) as mouse_activity_count
            from session_replay_events
            group by distinct_id -- this table is also always implicitly grouped by session id
            order by mouse_activity_count desc
            """
            ),
        )

        response = runner.calculate()
        assert response.results == [("distinct_id_two", 500), ("distinct_id_one", 100)]
