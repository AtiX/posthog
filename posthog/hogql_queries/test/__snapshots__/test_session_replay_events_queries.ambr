# serializer version: 1
# name: TestSessionReplayEventsHogQLQueries.test_session_replay_events_table_is_always_grouped_by_session_id
  '''
  SELECT session_replay_events.distinct_id AS distinct_id,
         sum(session_replay_events.mouse_activity_count) AS mouse_activity_count
  FROM
    (SELECT any(session_replay_events.distinct_id) AS distinct_id,
            sum(session_replay_events.mouse_activity_count) AS mouse_activity_count
     FROM session_replay_events
     WHERE equals(session_replay_events.team_id, 2)
     GROUP BY session_replay_events.session_id) AS session_replay_events
  GROUP BY session_replay_events.distinct_id
  ORDER BY mouse_activity_count DESC
  LIMIT 101
  OFFSET 0 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1
  '''
# ---
